File: code/494/Java/dp1.java
class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        return dp(nums, nums.length-1, S, new HashMap<>());
    }
    public int dp(int[] nums, int i, int target, Map<String, Integer> memo) {
        if (i==-1) {
            return target==0?1:0;
        }
        String key = String.valueOf(i)+','+String.valueOf(target);
        if (!memo.containsKey(key)) {
            memo.put(key, dp(nums, i-1, target+nums[i], memo)+dp(nums, i-1, target-nums[i], memo));
        }
        return memo.get(key);
    }
}

File: code/494/Python/dp1.py
class Solution:
    def findTargetSumWays(self, nums, S):
        def dp(i, target):
            if i==-1: return 1 if target==0 else 0
            if (i, target) not in memo:
                memo[(i, target)] = dp(i-1, target-nums[i])+dp(i-1, target+nums[i])
            return memo[(i, target)]
        return dp(len(nums)-1, S)